ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,264 @@
+// messageID => SubnetValidation
+mapping(bytes32 => SubnetValidation) subnetValidations;
+```
+
+#### About `validation`s
+
+A `validation` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator weight changes during the same validation.
+
+Each validation is identified by the `MessageID` corresponding to the Warp message in the `RegisterSubnetValidatorTx` that added the validator.
+
+`subnetValidations` will allow custom “security modules” (e.g. a PoS module) to take appropriate action upon node operator interactions (e.g. rewarding the node operator upon unstake). The `SubnetValidationPeriod` list will enable proper reconstruction of the validation timeline for each validator.
+
+### External functions
+
+The `ValidatorSetManager` contract should implement the following `external` functions.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// Returns the ID of the Subnet tied to this contract
+function getSubnetID() external view returns (bytes32);
+
+// Returns a Subnet validator information by its NodeID
+function getSubnetValidator(bytes32 nodeID) external view returns (SubnetValidator);
+
+/**
+ * Used to issue a RegisterSubnetValidatorTx
+ * @param nodeID The ID of the node to add to the Subnet
+ * @param weight The weight of the node on the Subnet
+ * @param expiry The time after which this message is invalid
+ * @param signature The Ed25519 signature of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[timestamp]
+ */
+function registerSubnetValidator(
+    bytes32 nodeID,
+    uint64 weight,
+    uint64 expiry,
+    bytes signature
+) external onlyOwner;
+
+/**
+ * Used to issue a SetSubnetValidatorWeightTx
+ * @param nodeID The ID of the node to modify
+ * @param weight The new weight of the node on the Subnet
+ */
+function setSubnetValidatorWeight(
+    bytes32 nodeID,
+    uint64 weight
+) external onlyOwner;
+
+/**
+ * Called to update the Subnet validator set upon receiving a notification Warp message from the P-Chain
+ * (after the successful addition or removal of a validator)
+ * @param messageIndex The index of the Warp message from the P-Chain
+ */
+function updateSubnetValidatorSet(
+    uint32 messageIndex
+) external;
+```
+
+#### About `ExitValidatorSetTx`
+
+This transaction allows a validator to exit a Subnet validator set directly from the P-Chain to prevent the Subnet from locking validators into participating in consensus indefinitely. Therefore it is not meant to be called in the `ValidatorSetManager` contract.
+
+### Information Needed from the P-Chain
+
+Upon successful addition, weight update, and removal of a validator from the Subnet, the P-Chain will send a notification in the form of a Warp message. The `updateSubnetValidatorSet` function is responsible for deserializing this message and updating the contract state accordingly.
```

Comment:
To be more clear here, the P-Chain is willing to sign a Warp message corresponding to validator set updates, but there still needs to be a delivery of that P-Chain message to the ValidatorManager to make it aware, and the delivery of the message is not part of the P-Chain's responsibility.